Blog post for image and distribution 1.1.0 (#152)
opencontainers/opencontainers.org: content/posts/blog/2024-03-13-image-and-distribution-1-1.md

---
title: OCI Image and Distribution Specs v1.1 Releases
author:
  name: Open Container Initiative
  tag: oci
tags: blog
date: 2024-03-13
---

The OCI Image Specification and Distribution Specification each had a 1.1.0 release on February 15, 2024.
These are the first minor releases since the 1.0.0 releases in July 2017 and May 2021 respectively, and include a number of notable changes.

> 

Reviewer: Add a "What's New" section? tl;dr and such
Basically, why are we doing this release? in two specs no less?

Reviewer: Also, include a "Clearing the Confusion" around previous dist-spec rc releases wrt "artifacts", since this is a blog post?

Reviewer: Also add a section about "Adopters"?
https://kubernetes.io/blog/2023/05/24/oci-security-profiles/
etc

Author: Thanks, details added.

The client requirement to manage the fallback tag is subject to race conditions since the existing content of the fallback tag must first be queried to add the new entry.
The fallback tags also clutter the tag listing with entries that are not the images end users are searching for.
Because of this, having registries upgrade to support the referrers API is preferred.

# Data Field

A `data` field was added to the definition of a [descriptor][image-descriptor] in the image specification.
This optional field contains the base64 encoded content of the blob.
When a client supports the field and encounters it, the additional requests to get the blob may be skipped.
Because many clients do not support this field, and base64 encoding adds overhead, it is recommended to only use this feature for small blobs.

# Manifest Maximum Size

When [pushing manifests][distribution-manifest-push], many clients and registries have size limits to prevent various resource exhaustion attacks.
The distribution specification now clarifies that a registry seeing its limit exceeded should return a 413 (payload too large) response to clients.

> And the specification also recommends that registries should support up to a 4MB image manifest (containing JSON descriptors, not the image layer content).

Reviewer 2: ```suggestion
And the specification also recommends that registries should support up to a 4MiB image manifest (containing JSON descriptors, not the image layer content).
```

Author: Thanks! I always mess that up. I should know better than to go by memory.

---
title: OCI Image and Distribution Specs v1.1 Releases
author:
  name: Open Container Initiative
  tag: oci
tags: blog
date: 2024-03-13
---

The OCI Image Specification and Distribution Specification each had a 1.1.0 release on February 15, 2024.
These are the first minor releases since the 1.0.0 releases in July 2017 and May 2021 respectively, and include a number of notable changes.

The biggest highlight of these releases is the process for associating metadata with existing images using a new `subject` field in the image specification and the associated referrers API in the distribution specification.
However, this is far from the only thing that was changed.
And many of these changes may affect existing OCI conformant projects.

# Guidance for Artifacts

> OCI has been used to package objects other than container images for some time, including projects like Helm, OPA, and WASM.

Reviewer: Nit ...

s/WASM/Wasm/ (per https://webassembly.org/)

Author: TIL. Thanks!

The guidance for how projects should implement artifacts has now been [codified into the image specification][image-artifact-guidance].

The recommendation to use a `config.mediaType` value different from the recognized container image values remains.
That guidance is now extended to include a new `artifactType` field that enables use cases without a dedicated `config.mediaType`.

> In the scenarios were an artifact does not have content for the config or any layers, an empty JSON descriptor is recommended as a placeholder to satisfy the requirement that these fields exist, while serving as an indication to consuming tools that there is no useful content in that descriptor.

Reviewer: a link to the "empty JSON descriptor"? ... an important enough constant to call out?

Reviewer: https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidance-for-an-empty-descriptor

Author: Added, good call out!

    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:e40ce69ad428e47eb72ea76d9...",
      "size": 701,
      "annotations": {
        "org.opencontainers.image.created": "2024-03-03T15:15:15Z",
        "org.opencontainers.image.description": "Another example artifact for same image"
      },
      "artifactType": "application/example2"
    }
  ]
}
```

# Support For Existing Registries

The design of the changes to the image and distribution specifications assume that it will take time for registries to upgrade.
To facilitate this transition period, changes to the image spec added new fields which should be ignored by registries without generating errors or modifying the content of the manifest.
Clients pushing a manifest with a `subject` field and querying for referrers have an additional responsibility to manage and use a [fallback tagged Index when the registry does not support the referrers API][distribution-referrers-fallback].
When a registry supports the referrers API, it will respond to pushing a manifest with the `subject` with the header `OCI-Subject` set to the digest in the subject field.

> When querying for referrers, a registry will respond with a `404` or similar error code if the API is not supported (when no referrers are found, an Index with an empty manifest list would be returned).

Reviewer: Language a little confusing ... are we mixing the cases of non-support and not-found?

Author: I did some rewording of these last two lines. Let me know if that helps.

---
title: OCI Image and Distribution Specs v1.1 Releases
author:
  name: Open Container Initiative
  tag: oci
tags: blog
date: 2024-03-13
---

The OCI Image Specification and Distribution Specification each had a 1.1.0 release on February 15, 2024.
These are the first minor releases since the 1.0.0 releases in July 2017 and May 2021 respectively, and include a number of notable changes.

> The biggest highlight of these releases is the process for associating metadata with existing images using a new `subject` field in the image specification and the associated referrers API in the distribution specification.

Reviewer 3: ```suggestion
# Artifacts

The biggest highlight for 1.1 is support for associating metadata artifacts with existing OCI images. New `subject` and `artifactType` fields are added to the image specification and retrieval of these associated artifacts can be done via a new referrers API in the distribution specification.
```

Author: Suggestions have been applied.

However, this is far from the only thing that was changed.
And many of these changes may affect existing OCI conformant projects.

> # Guidance for Artifacts

Reviewer 3: ```suggestion
## Guidance for Artifacts
```

OCI has been used to package objects other than container images for some time, including projects like Helm, OPA, and Wasm.
The guidance for how projects should implement artifacts has now been [codified into the image specification][image-artifact-guidance].

The recommendation to use a `config.mediaType` value different from the recognized container image values remains.
That guidance is now extended to include a new `artifactType` field that enables use cases without a dedicated `config.mediaType`.
In the scenarios were an artifact does not have content for the config or any layers, an [empty JSON descriptor is recommended][image-empty-descriptor] as a placeholder to satisfy the requirement that these fields exist, while serving as an indication to consuming tools that there is no useful content in that descriptor.

> # Describing Associations

Reviewer 3: ```suggestion
## Describing Associations
```

This is used by the [Referrers API](#referrers-api) described below.
It allows tools to associate signatures, attestations, and other metadata with images with a common specification.

Putting together the advice for packaging artifacts with the `subject` field to associate the artifact with an image, results in a manifest that may look like:

```jsonc
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/example",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146a...",
    "size": 2
  },
  "layers": [
    {
      "mediaType": "application/example",
      "digest": "sha256:e3b0c44298fc1c149af...",
      "size": 0
    }
  ],
  "subject": {
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "digest": "sha256:21edd7d11800e94bae9f4...",
    "size": 1769
  },
  "annotations": {
    "org.opencontainers.image.created": "2024-03-02T15:16:17Z",
    "org.opencontainers.image.description": "Example artifact for image sha256:21ed..."
  }
}
```

> # Referrers API

Reviewer 3: ```suggestion
## Referrers API
```

Including those values allows tooling to select their desired artifact from the OCI Index the same way runtimes use the platform to select an image to run from a multi-platform Index.

The referrers response to `GET /v2/<name>/referrers/sha256:21edd7d11800e94bae9f4...` may look like:

```jsonc
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:9e6569b15e5ed981334003eb8...",
      "size": 724,
      "annotations": {
        "org.opencontainers.image.created": "2024-03-02T15:16:17Z",
        "org.opencontainers.image.description": "Example artifact for image sha256:21ed..."
      },
      "artifactType": "application/example"
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:e40ce69ad428e47eb72ea76d9...",
      "size": 701,
      "annotations": {
        "org.opencontainers.image.created": "2024-03-03T15:15:15Z",
        "org.opencontainers.image.description": "Another example artifact for same image"
      },
      "artifactType": "application/example2"
    }
  ]
}
```

> # Support For Existing Registries

Reviewer 3: ```suggestion
## Support For Existing Registries
```

The design of the changes to the image and distribution specifications assume that it will take time for registries to upgrade.
To facilitate this transition period, changes to the image spec added new fields which should be ignored by registries without generating errors or modifying the content of the manifest.
Clients pushing a manifest with a `subject` field and querying for referrers have an additional responsibility to manage and use a [fallback tagged Index when the registry does not support the referrers API][distribution-referrers-fallback].
When a registry supports the referrers API, it will return the header `OCI-Subject` set to the digest in the `subject` field in response to a client pushing a manifest with the `subject` field.

> Similarly, when querying for referrers, a registry that supports the API will always return a `200` response, even if the response is an empty Index, since registries that have not implemented the API may respond with a `404` or similar error.

Reviewer: ", since registries" ... 

"to distinguish between registries" ...

Author: Added, though I think it makes the rest of the sentence a bit harder to read.

Reviewer: What we want to say I suppose is 2xx and 4xx status codes are used to tell if the referrer api is supported or not.

The syntax for an extension API is either `/v2/_<extension>/<component>/<module>[?params]` for registry level extensions, or `/v2/<name>/_<extension>/<component>/<module>[?params]` for repository level extensions.
For example, the following extension is used to list installed extensions: `GET /v2/<name>/_oci/ext/discover`.
To prevent conflicts, extensions should register their name with the distribution specification.

# Multiple Matching Platforms

When multiple entries in an [OCI Index][image-index] appear to be the same, runtime tooling should prefer the first match.
This clarification allows future extensibility for tooling that may wish to add additional images for the same platform that would be detected and used by their tooling.

To get involved with efforts to increase the platform extensibility, see the [OCI Working Group on Image Compatibility][wg-image-compatibility].

# Warning Header

Registries may now include a [warning header][distribution-warning] in responses.
This can be used to notify users of issues that are not yet error conditions, including deprecation announcements.

# No Artifact Manifest

> Notably absent from this release is a new manifest dedicated to artifacts.

Reviewer 3: ```suggestion
Notably absent from this release is a new type of manifest dedicated to artifacts.
```
